lib/charms/grafana_k8s/v0/grafana_source.py

```diff
@@ -0,0 +1,433 @@
+            version: a string providing the semantic version of the Grafana
+                source being provided.
+
+        """
+        super().__init__(charm, name, service, version)
+        self.charm = charm
+        events = self.charm.on[name]
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
+
+        self.framework.observe(
+            events.relation_changed, self.on_grafana_source_relation_changed
+        )
+        self.framework.observe(
+            events.relation_broken, self.on_grafana_source_relation_broken
+        )
+
+    def on_grafana_source_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle relation changes in related consumers.
+
+        If there are changes in relations between Grafana source providers
+        and consumers, this event handler (if the unit is the leader) will
+        get data for an incoming grafana-source relation through a
+        :class:`GrafanaSourcesChanged` event, and make the relation data
+        is available in the app's datastore object. The Grafana charm can
+        then respond to the event to update its configuration
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        rel = event.relation
+        data = (
+            json.loads(event.relation.data[event.app].get("sources", {}))
+            if event.relation.data[event.app].get("sources", {})
+            else None
+        )
+        if not data:
+            logger.info("NO DATA")
+            return
+
+        # This is due to a double fire on relation_joined calling
+        # the consumer, which calls relation_changed, and the actual data may be
+        # the same, but we'll see the source name in use later, try to set it again,
+        # and get double relation names
+        if (
+            self._stored.sources.get(rel.id, None)
+            and len(set(data.items()) ^ set(self._stored.sources[rel.id].items())) == 0
+        ):
+            logger.debug("Relation data unchanged. Doing nothing")
+            return
+
+        source = SourceData(event.app.name, event.unit, event.app, rel.id, data)
+        try:
+            data = _validate(self, source)
+        except SourceFieldsMissingError as e:
+            logger.critical(
+                f"Missing data on added grafana-k8s source {e}", exc_info=True
+            )
+            return
+
+        logger.info("ADDING SOURCE: {}".format(data))
+
+        self._stored.sources[rel.id] = data
+        self.on.sources_changed.emit()
+
+    def on_grafana_source_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Update job config when consumers depart.
+
+        When a Grafana source consumer departs, the configuration
+        for that consumer is removed from the list of sources jobs,
+        added to a list of sources to remove, and other consumers
```

Comment:
Since this is the Provider class, shouldn't this read "and the provider is informed through a" rather than "and other consumers are informed through a" ?

Reply:
Typo. I'll resolve.